### PR TITLE
K8s merge develop

### PR DESCRIPTION
**Pull request by wallyworld**
## Description of change

Merge develop to primarily bring in the new k8s labels work in #12065
This is in https://github.com/juju/juju/pull/12071/commits/2b3cafaf7377c1a0b55438d1029efe2b003208d9

## QA steps

bootstrap microk8s
deploy a charm
inspect the k8s resources
remove an application

### CODE REVIEW

**Comment by wallyworld** [comment]
!!build!!

File: caas/kubernetes/provider/application/application.go [hunk 1]
```diff
@@ -875,18 +885,23 @@ func (a *app) applicationPodSpec(config caas.ApplicationConfig) (*corev1.PodSpec
...
+		Merge(k8sutils.AnnotationsForVersion(config.AgentVersion.String(), a.legacyLabels))
 }
 
 func (a *app) labels() map[string]string {
 	// TODO: add modelUUID for global resources?
-	return map[string]string{constants.LabelApplication: a.name}
+	return k8sutils.LabelsForApp(a.name, a.legacyLabels)
+}
+
+func (a *app) selectorLabels() map[string]string {
```
Comment (tlm) [inline caas/kubernetes/provider/application/application.go:896]: It may be worth changing this function signature to return labels.Set

**Comment by wallyworld** [comment]
[devel / k8s-spike merge](https://trello.com/c/7lGq9xml/1198-devel-k8s-spike-merge)

**Review by tlm — APPROVED** [review]
lgtm, monster merge. Thanks

File: caas/kubernetes/provider/application/application.go [hunk 2]
```diff
@@ -875,18 +885,23 @@ func (a *app) applicationPodSpec(config caas.ApplicationConfig) (*corev1.PodSpec
 
 func (a *app) annotations(config caas.ApplicationConfig) annotations.Annotation {
 	return k8sutils.ResourceTagsToAnnotations(config.ResourceTags).
-		Add(constants.LabelVersion, config.AgentVersion.String())
+		Merge(k8sutils.AnnotationsForVersion(config.AgentVersion.String(), a.legacyLabels))
 }
 
 func (a *app) labels() map[string]string {
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply caas/kubernetes/provider/application/application.go]: done

File: caas/kubernetes/provider/application/application.go (hunk 1 above)
Comment (wallyworld) [reply caas/kubernetes/provider/application/application.go]: done

**Comment by wallyworld** [comment]
$$merge$$

**Comment by wallyworld** [comment]
$$merge$$

**Comment by wallyworld** [comment]
$$merge$$

**Comment by wallyworld** [comment]
$$merge$$

**Pull request merged**